File: 240616/정수 사각형 최댓값의 최소/minimax-path-in-square.py
n= int(input())
board = [list(map(int, input().split())) for _ in range(n)]

dp = [[0] * n for _ in range(n)]

dp[0][0] = board[0][0]

for x in range(1, n):
    dp[0][x] = max(dp[0][x-1], board[0][x])

for y in range(1, n):
    dp[y][0] = max(dp[y-1][0], board[y][0])

for y in range(1, n):
    for x in range(1, n):
        dp[y][x] = max(board[y][x], min(dp[y-1][x], dp[y][x-1]))

print(dp[n-1][n-1])

File: 241014/정수 사각형 최댓값의 최소/minimax-path-in-square.java
import java.util.*;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int N = sc.nextInt();
        int[][] board = new int[N][N];
        for(int i = 0; i < N; i++){
            for(int j = 0; j < N; j++){
                board[i][j] = sc.nextInt();
            }
        }
        int[][] dp = new int[N+1][N+1];
        dp[1][1] = board[0][0];
        Arrays.fill(dp[0], Integer.MAX_VALUE);
        for(int i = 1; i <= N; i++){
            dp[i][0] = Integer.MAX_VALUE;
        }
        for(int i = 1; i <= N; i++){
            for(int j = 1; j <= N; j++){
                if(i == 1 && j== 1) continue;
                int other = Math.min(dp[i-1][j], dp[i][j-1]);
                dp[i][j] = Math.max(other, board[i-1][j-1]);
            }
        }

        System.out.println(dp[N][N]);
    }
}